library/src/schemas/enum/enum.ts

```diff
@@ -18,6 +18,31 @@ export interface Enum {
   [key: string]: string | number;
 }
 
+type IsNumericString<K extends string> = K extends
+  | 'NaN'
+  | 'Infinity'
+  | '-Infinity'
+  ? true
+  : K extends `${infer V extends number}`
+    ? `${V}` extends K
+      ? true
+      : false
+    : false;
+/**
+ * enum_ only accepts enum values of string & non-numeric keys.
+ * This type-level function filters out numeric keys including Infinity and NaN.
+ *
+ * @example FilterEnumKeys<'foo' | '1' | 'Infinity' | 'NaN'> will be 'foo'.
+ */
+type FilterEnumKeys<K extends string> = K extends K
+  ? IsNumericString<K> extends true
+    ? never
+    : K
+  : never;
+type EnumValues<TEnum extends Enum> = TEnum[FilterEnumKeys<
```

Comment:
I did not fully understand why `FilterEnumKeys` is needed. Can you provide a minimal example via our [playground](https://valibot.dev/playground/) that demonstrates that the current implementation results in a wrong type?

Reply:
These are my motivating examples:

https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6BjCAO1XgGUmALAUxDI4AXkwA6PqwCuIABQBvOAElW2YK2AwAngC44ABjgBfAJQNmbDnCh9UU0vFEYxqZNj4AFZFFR9Z3fkEAGgMzRhZ2CFI+MVIIAHNZGzsHMzgAenTlVhJyABM4bTA+PQBRAA9iphg+AtY+DD4oOAopeBsmPmBGgv1GWQB9PX0TPWcVdygAeTawNoAeIr4IbDgAgWQAPhFtgdoMrJmYOfhUHgh7AvUmUik8vgN6IA
```ts
import * as v from 'valibot';

const Schema = v.enum({ Infinity: 0 });

const result = v.safeParse(Schema, 0);

console.log(result); // Invalid type: Expected never but received 0

(_: 0): v.InferOutput<typeof Schema> => _; // Output should include 0
```

https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6BTAOwFcCBBOAb3rjgBpOAF44AOWTj6AX0YBjCC1TwAyvIAWTEMlGYAdKw4AKTgEoG9RcvhQmqNqXhiM+1MmxMACsiiomxupaOgA0RIKEFgpKqBCkTPqkEADmxnYOThZwAPTZcA7y8vaojMYA+gBc4ZFVrgCSLJ5QAPJsMGBtADwwAJ5gTBDYcEHayAB8ohNltDl5re1t+RoQjgAmcCzUcMAs8qRsq0zV9EA
```ts
import * as v from 'valibot';

enum A {
  K = NaN
}

const Schema = v.enum(A);

const result = v.safeParse(Schema, 'K');

console.log(result); // success

(_: 'K'): v.InferOutput<typeof Schema> => _; // Output should not include 'K'
```

https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6BTAOwFcCBBOAb3riICAjADpBAXiIB5Fk0L0AvowDGEFqngBlZQAsmIZHEkYRrDgApOASgb1V6+FCao2peMZGpk2JgAVkUKhM5tp6BgA00rKENipqqBCkTCKkEADm5k4ubjZwAPR5cACSLCTkACZwMACeYEwAXHAAogAedcowTJWyGExQcBRsjkzKTMC9lQA6AEQyTDOM5gD6jeY1dRDYcNYA2oTCYoQAulaNJiU+UFJDYEMAPOtMm3Ch+sgAfEafS7T5hdcwW7wVA6CCuSrAFjKUhscpMKJyIA
```ts
import * as v from 'valibot';

enum A {
  '  1. ' = 'One'
}

const Schema = v.enum(A);

const result = v.safeParse(Schema, 'One');

console.log(result); // Invalid type: Expected never but received \"One\"

(_: (typeof A)['  1. ']): v.InferOutput<typeof Schema> => _; // Output should include 'One'
```

Reply:
Thanks for the code examples. I'm very sorry, but I still don't understand why `FilterEnumKeys` is necessary.  It is true that there is currently a bug in the way we filter the enum options, which leads to wrong runtime behavior, but the type information should be correct.

Reply:
@fabian-hiller A wrong type occurs when enum_ accepts an object literal and the literal contains reverse mapping entries.

https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6BjCAO1XgGUmALAUxDI4AXkwA6PqwCuIABQBvenGVwAWnzwAuOAAYANEpU7thdXkIGVcAIwmA8qz4X6AXwCUDGAE8wfON35BOykYMBCRcQBJVmwNYNCQgB5vXwhsf14BZAA+WjgAenzdOAAfOAAiMwhy0oqHPnL6QvFJGRxgUhgNdAhwoQBrPi9dPThUCAzA5Hiw+FQeXtIAEzgKPx1a8vrGoA
```
import * as v from 'valibot';

const Schema = v.enum({
    Zero: 0,
    0: 'Zero',
    1: 'One',
});
type SchemaOutput = v.InferOutput<typeof Schema>; // 0 | "Zero" | "One"
// v.enum filters out a key 0, so SchemaOutput should be 0 | "One"
```

Reply:
In a perfect world, I would only filter reverse mapping entries of "real" TS enums and not of normal objects, as this could lead to unexpected behaviour if someone accidentally provided an object with a reverse mapping entry that was not intended as a reverse mapping entry. But the problem is that we can't tell at runtime what is a TS enum and what is a normal object. Therefore, I will probably follow your recommendations and document this behavior in our API reference.